importlib_resources/abc.py

```diff
@@ -108,9 +112,15 @@ def joinpath(self, *descendants: StrPath) -> "Traversable":
             path.parts for path in map(pathlib.PurePosixPath, descendants)
         )
         target = next(names)
-        return next(
+        matches = (
             traversable for traversable in self.iterdir() if traversable.name == target
-        ).joinpath(*names)
+        )
+        try:
+            return next(matches).joinpath(*names)
+        except StopIteration:
+            raise TraversalError(
```

Comment:
The only way for this to happen is no arguments being passed, no? Wouldn't checking for that directly make the code easier to read?

Reply:
I don't think that's correct. A `StopIteration` would occur if none of the `traversable.name == target` in the `matches` above (including if `self.iterdir()` yields no results). It's trapping `StopIteration` on `next(matches)`, not `*names`. `*names` can be empty. Maybe only the exception should be trapped for `next(matches)`.

Reply:
Ah, right, that makes sense.